docs/arcadia-users-group/20221031-binder/lesson.md

```diff
@@ -0,0 +1,171 @@
+   > **https://mybinder.org/v2/gh/YOUR-USERNAME/2022-my-py-binder/HEAD**
+4. Copy it, open a new browser tab and visit that URL
+   - You will see a "spinner" as Binder launches the repo
+
+If everything ran smoothly, you'll see a JupyterLab interface.
+
+**What happens when you launch a Binder?**
+In the background, BinderHub (the backend of Binder) fetches your repo from GitHub and analyzes the contents of the files in the repo.
+It then builds a docker image based on your repo and launches the docker image in the cloud.
+Lastly, BinderHub connects you to the instance running the docker image via your browser.
+
+## Interacting with files in Binder
+
+The Binder environment is fully executable. 
+Let's execute our python script from the command line.
+Click the `Terminal` button to open the Terminal app, then run:
+
+```
+python hello.py
+```
+
+`Hello from Binder!` should be printed to the terminal.
+
+## Building out the environment and pinning dependencies
+
+By including a `.py` file in our repository, BinderHub knew that we wanted to launch a Binder with python installed.
+We can provide more explicit instructions to customize the installation experience.
+We'll do this using conda because conda allows us to install python, R, ... dependencies.
+
+The `environment.yml` is the standard configuration file used by conda that lets you install any kind of package, including Python, R, and C/C++ packages.
+BinderHub will use this file to update the `base` conda environment with the packages listed in your `environment.yml` -- it does _not_ create and activate a new conda environment. 
+This means that the environment will always have the same default name, not the name specified in your `environment.yml`.
+
+You can install files from pip in your environment.yml as well. 
+For example, see the [binder-examples `environment.yml` file](https://github.com/binder-examples/python-conda_pip/blob/HEAD/environment.yml).
+
+We'll start by creating an `environment.yml` and pinning a dependency.
+
+* In your repo, create a file called `environment.yml`
+* Add the following lines:
+```
+channels:
+  - conda-forge
+  - bioconda
+  - defaults
+dependencies:
+  - numpy=1.14.5
+```
+* Check for typos! Then commit to the main branch
+* Visit <https://mybinder.org/v2/gh/YOUR-USERNAME/2022-my-py-binder/HEAD> again in a new tab
+
+This time, BinderHub will read the configuration file you added and install the specific version of the package you requested using conda.
+
+## Using dependencies in the notebook environment
+
+1. From the launch panel, select "Python 3" from the Notebook section to open a new notebook
+2. Type the following into a new cell and run it with <kbd>Shift</kbd>+<kbd>Enter</kbd> 
+```
+import numpy
+print(numpy.__version__)
+numpy.random.randn()
+```
+
+> **If you save this notebook, it will not be saved to the GitHub repo.** 
+Pushing changes back to the GitHub repo through the container is not possible with Binder. 
```

Comment:
Why is this not possible? I thought it came with git installed? Even so, I'd emphasize maybe two things:
* You can still download files from Binder
* It's not suitable for making significant changes to code

Reply:
k! i'll emphasize these things!

This line was from the turing training...and i just tested it and it's not true. You have to go through the whole create-a-key-and-configure-ssh rigamarole, but you can push back if you have write access to the repo. anyway, i'll change this section!